File: Leet-Code/maximum_length_of_good_sequence.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        n = len(nums)
        if n == 0:
            return 0

        # Initialize the dp array where dp[i][j] represents the maximum length of a good subsequence
        # ending at index i with exactly j changes.
        dp = [[0] * (k + 1) for _ in range(n)]

        # Initialize dp[i][0] to 1 for all i since a single element is always a valid subsequence with 0 changes
        for i in range(n):
            dp[i][0] = 1

        # Fill the DP table
        for i in range(1, n):
            for j in range(i):
                for changes in range(k + 1):
                    if nums[i] == nums[j]:
                        dp[i][changes] = max(dp[i][changes], dp[j][changes] + 1)
                    elif changes > 0:
                        dp[i][changes] = max(dp[i][changes], dp[j][changes - 1] + 1)

        # Get the maximum value from the dp array considering up to k changes
        max_length = 0
        for i in range(n):
            for changes in range(k + 1):
                max_length = max(max_length, dp[i][changes])

        return max_length

File: Leet-Code/maximum_length_of_good_sequence.cpp
class Solution {
public:
    std::map<int, int> f[55];
    int n, k, g[55], a[5005];
    int maximumLength(vector<int>& nums, int k_) {
        n = nums.size();
        k = k_;
        for(int i = 1; i <= n; ++i) a[i] = nums[i - 1];
        for(int i = 1; i <= n; ++i){
            for(int s = 0; s <= k; ++s){
                f[s][a[i]] = f[s][a[i]] + 1;
                if(s > 0) f[s][a[i]] = std::max(f[s][a[i]], g[s - 1] + 1);
            }
            for(int s = 0; s <= k; ++s) g[s] = std::max(g[s], f[s][a[i]]);
        }
        int ans = 0;
        for(int i = 0; i <= k; ++i) ans = std::max(ans, g[i]);
        return ans;
    }
};